Allow setting Postgres version via build-time argument

LeoVerto opened the pull request:
This PR allows setting the postgres version in it's dockerfile by using the `POSTGRES_VERSION` build-time argument (note `PG_VERSION` would collide with env vars of the parent image). The default is set to 9.5 as to not break workflows of users using the image directly.

It also makes use of the arg in the compose file where the version can easily be changed. Using the `image` option is necessary to force rebuilds on arg changes and makes switching between versions easier.

While this setup is very backwards-compatible, updating the postgres version would require changing it in all three places (although changing the Dockerfile is not strictly required). A build script or .env file could further optimize this.

LeoVerto committed:
Allow setting Postgres version via build-time argument

mayhem reviewed (commented):
Looks ok to me, but I'm guessing that somewhere the postgres client gets installed as well -- that would need to have the matching version installed, no?

LeoVerto commented:
If you mean the postgres client inside the container: the correct version is already included in the `postgres` base version:
```
$ docker run -it --rm musicbrainz-docker_db:10 sh
# psql --version
psql (PostgreSQL) 10.6 (Debian 10.6-1.pgdg90+1).
```

mayhem commented:
OK, great. Thanks!

yvanzo reviewed (changes requested):
Would be nice to allow setting PostgreSQL version from `docker-compose` command-line without having to hack into the compose file. Also, can you please document this option into the README.md doc?

yvanzo committed:
Tag db image according to POSTGRES_VERSION

Image cannot be tagged after build arg defined on the above line,
thus it must be defined with default value and may be set from
either build argument or environement variable (set either in .env
file or by command-line parameter).  To ensure consistency between
these two lines, build arg follows environement variable too.

yvanzo commented:
Potential issue to look into: `psql` is installed in `musicbrainz` service too. It is used for maintenance and development mostly.

Edit: issue confirmed, additionally to Perl module `DBD::Pg` build depending on `libpq-dev`. Both are now addressed with https://github.com/metabrainz/musicbrainz-docker/pull/89/commits/7c9e6eb3098dab3f72cc332e4b028a2b89e12fe5.

yvanzo commented:
Update: migration to SIR will be made prior to address this issue, as SIR depends on PostgreSQL too.

yvanzo committed:
Build musicbrainz according to POSTGRES_VERSION

- Set build argument POSTGRES_VERSION from environment variable;
- Add PGDG APT repository https://wiki.postgresql.org/wiki/Apt
  which holds packages for versions 9.x, 10, and 11 of postgres;
- Set default value for POSTGRES_VERSION build argument to 9.5;
- Install 'postgres-${POSTGRES_VERSION}' package (required for
  maintenance/admin scripts through `admin/psql`) and keep
  installing `libpq-dev` which provides `pg_config` as a build
  dependency to install the Perl module DBD::Pg.

yvanzo committed:
Document setting the env. var. POSTGRES_VERSION

yvanzo commented:
@LeoVerto: After all, SIR still requires more work before being shipped in there. Thus I rebased your PR on master and completed it with commits for: setting Postgres version just from environment, rebuilding `musicbrainz` service with the same version of Postgres, and adding a line in documentation. Please take a look at it, thanks!

LeoVerto commented:
Looks good to me, I didn't even know you could set default values for env vars in compose files like that!

I noticed you added the postgres package installation after the MBS one, does this mean that the MBS version is going to change more rarely than the PG one? Otherwise the other way around might be better for intermediate layer re-usability.

yvanzo commented:
Good catch, I focused on development use case but for mirroring you’re too right! Fixed in [`80a421a`](https://github.com/metabrainz/musicbrainz-docker/commit/80a421ab66e85847a521b732a4abecd8e9f45b1f#diff-906e5d52e79584e237931f661ae3ab8a).

yvanzo commented:
Thank you @LeoVerto !
